project/SUNGOD3/README.rst

```diff
@@ -0,0 +1,88 @@
+Advanced Vector Search for RAG Systems
+======================================
+
+Basic Information
+-----------------
+This project, tentatively named "AdvancedVectorSearch", aims to develop and implement novel vector search methods for RAG systems.
+
+Vector search is a crucial component in modern information retrieval systems, particularly in RAG applications. While FAISS is a well-established library for efficient similarity search, there's room for innovation in search methodologies. 
+
+Note that the goal of this project will focus on vector search, similarity comparison function, vector data storage, and does not include matrix calculations of LLM.
+
+Problem to Solve
+----------------
+The primary challenges this project aims to address are:
+
+1. Improving search speed beyond basic linear search, especially for high-dimensional vector spaces common in RAG systems.
+2. Developing new similarity metrics that may be more suitable for specific RAG applications.
+3. Creating a flexible, extensible framework that allows for easy experimentation with different search algorithms and metrics.
+4. Balancing speed, accuracy, and memory usage in vector search operations.
+
+Prospective Users
+-----------------
+This project will benefit:
+
+1. Researchers and developers working on RAG systems.
+2. Data scientists and machine learning engineers dealing with large-scale similarity search problems.
+3. Information retrieval specialists looking for customizable vector search solutions.
+4. Anyone working with high-dimensional data requiring efficient similarity search capabilities.
+5. Graduate students who lack graduation thesis.
+
+System Architecture
+-------------------
+The AdvancedVectorSearch system will consist of the following components:
+
+1. Dataset Module: For generating and managing large-scale, high-dimensional vector datasets for testing and benchmarking.
+2. Index Module: Implementing various similarity comparison functions and indexing structures.
+3. Search Module: Providing add and search functionalities with support for different search algorithms.
+4. Benchmark Module: For comparing the implemented methods against linear search and potentially FAISS in terms of speed, accuracy, and memory usage.
+
+The system will be implemented in C++ for core functionalities, with Python bindings for ease of use.
```

Comment:
Do you refer any paper or book to implement your architecture? I'm wondering how to do that.

Reply:
I didn't specifically refer to any paper to design this architecture. The modular structure is based on my understanding of RAG. However, your comment reminds me that I should add the RAG paper as a reference in the documentation.
Regarding implementation, I designed these modules to help abstract similar functionalities (for easier comparison and reduced dependencies). I believe that implementing a linear search RAG without any parallelization isn't particularly challenging. In fact, I have experience implementing this process in Python (without using libraries) during a hackathon. Rewriting it in C++ shouldn't be a problem. For the FAISS part, I plan to study their examples and API for direct use. The more challenging aspect would be developing my own similarity search algorithm/index, for which I hope to draw inspiration from our classes.